eth: stm32h7: use SRAM3 / MPU configuration

mjaun opened the pull request:
Fixes: #29915

Changes are based on [these recommendations](https://community.st.com/s/article/FAQ-Ethernet-not-working-on-STM32H7x3).

Notes:

* Putting the descriptors in a separate MPU region and configuring them as device memory is what actually fixed the issue. I am not sure why ST recommends using SRAM3, perhaps there are performance reasons but maybe also other potential issues.
* The Cortex M linker script should not contain STM32 specific definitions. There should be a more generic way to define the required sections.
* Same applies for the static MPU configuration. Ideally both could be done in the device tree as proposed by @Nukersson in the issue.
* There is already a PR with another approach: #29944

Nukersson commented:
Thank you very much for this very useful PR.

Now my ideas on it:

1. @emillindq Already provided solution and test with ST's driver [fix](https://github.com/STMicroelectronics/STM32CubeH7/issues/95). My suggestion here would be push it anyway, cause it does only __DSB() after writing eth descriptors and therefore should not have impact on ethernet throughout or driver performance.

2.@reloZid Very elegant fix from you here! But we can't just change linker scripts every time we need specific memory regions configuration. I think here we need memory reallocation and MPU configuration @possibilities in DT

@gmarull @erwango @galak Now we seems to have proper solution for solide big fix with future improvents for the whole Zephyr project. What do you think? Especially, what do you think about MPU configuration and data regions definition in DT?

gmarull commented:
Thanks for this draft, will be useful to trigger discussions on how to find a proper fix for this problem and also future similar requirements.

I see some other projects have a sort of MPU API, e.g. https://github.com/u-boot/u-boot/blob/master/arch/arm/mach-stm32/soc.c

In Zephyr I think something similar can be done if using `CONFIG_CPU_HAS_CUSTOM_FIXED_SOC_MPU_REGIONS` (see NXP k6x/k8x for example) and provide a soc level MPU configuration. In this case, as ST recommends SRAM3 for I/O Ethernet/USB:

![image](https://user-images.githubusercontent.com/25011557/100938208-94ea9c00-34f4-11eb-940a-fe2f2ea760ea.png)

SRAM3 could be configured entirely for that purpose at soc level, and just provide DT definitions for sram3 together with linker scripts facilities so that Ethernet buffers can easily be placed there. I'm not sure if that's the best approach, though.

I'm not really familiar with Zephyr MPU implementation, it would be good to have input from maintainers @ioannisg @galak @MaureenHelm @stephanosio @carlocaione

Nukersson commented:
We need a possibility to configure MPU and SRAM sections over DT or KConfig for Zephyr. And then we need a possibility to use configured SRAM sections in our code.

galak commented:
dev-review (Dec 3, 2020): 

Suggestion to utilize CONFIG_NOCACHE_MEMORY and look at adding one more section for the SRAM specific case in arch/arm/core/aarch32/cortex_m/mpu/arm_core_mpu.c.

@ioannisg would like to see over time cleaning up of the linker script and mpu tables to move into SoC specific code.

Nukersson commented:
> dev-review (Dec 3, 2020):
> 
> Suggestion to utilize CONFIG_NOCACHE_MEMORY and look at adding one more section for the SRAM specific case in arch/arm/core/aarch32/cortex_m/mpu/arm_core_mpu.c.

I would prefer not to add _half-solutions_. We can better define amount of work and split it here, providing proper and solid solution.

> 
> @ioannisg would like to see over time cleaning up of the linker script and mpu tables to move into SoC specific code.

Nukersson commented:
> dev-review (Dec 3, 2020):
> 
> Suggestion to utilize CONFIG_NOCACHE_MEMORY and look at adding one more section for the SRAM specific case in arch/arm/core/aarch32/cortex_m/mpu/arm_core_mpu.c.

I would prefer not to add _half-solutions_. We can better define amount of work and split it here, providing proper and solid solution.

> 
> @ioannisg would like to see over time cleaning up of the linker script and mpu tables to move into SoC specific code.

KozhinovAlexander reviewed (changes requested):
The idea behind the changes is very good. But we need more scalable solution.

erwango commented:
> The idea behind the changes is very good. But we need more scalable solution.

@Nukersson, this point was discussed during dev-review meeting. Of course this proposal lacks scalability. Though w/o alternative proposal we find it acceptable for now with few tweaks.
To define a more scalable alternative we'd need more variety of usage examples. A rework of the whole way to instantiate memory partitions from DT is something we'd like to get, but it has a lot of intrications so it was proposed to postpone post LTS (not before 6 monthes).

Nukersson commented:
> > The idea behind the changes is very good. But we need more scalable solution.
> 
> @Nukersson, this point was discussed during dev-review meeting. Of course this proposal lacks scalability. Though w/o alternative proposal we find it acceptable for now with few tweaks.
> To define a more scalable alternative we'd need more variety of usage examples. A rework of the whole way to instantiate memory partitions from DT is something we'd like to get, but it has a lot of intrications so it was proposed to postpone post LTS (not before 6 monthes).

Okay. I understand it. But then we need to plan proper solution within the roadmap for next year or two.

ioannisg commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -106,6 +106,9 @@ MEMORY
     SRAM1                 (rw) : ORIGIN = RAM1_ADDR, LENGTH = RAM1_SIZE
     SRAM2                 (rw) : ORIGIN = RAM2_ADDR, LENGTH = RAM2_SIZE
 #endif
+#ifdef CONFIG_ETH_STM32_HAL_USE_SRAM3
```
The MEMORY description here already contains several other SoC-specific sections so having one more does not do any additional harm.

I would like to see this cleaned up in the future, and keep all SoC specific section in the corresponding SoC definitions.

ioannisg replied:
However, could we at least align with the more modern concept of doing this via Device Tree and not Kconfig?

reloZid replied:
Done.

ioannisg commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -446,6 +449,35 @@ GROUP_START(DTCM)
 GROUP_END(DTCM)
 #endif
 
+#ifdef CONFIG_ETH_STM32_HAL_USE_SRAM3
```
This could also be an #ifdef directive based on DT.

reloZid replied:
Done.

ioannisg commented on soc/arm/st_stm32/stm32h7/linker.ld:
```diff
@@ -6,4 +6,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#define RAM3_SIZE (64 * 1K)
```
This won't be needed, imho, if we use device tree for getting the RAM3 address and size.

reloZid replied:
Done.

ioannisg commented on include/arch/arm/aarch32/cortex_m/mpu/arm_mpu_v7m.h:
```diff
@@ -262,6 +262,10 @@ typedef struct {
 	{(_K_MEM_PARTITION_P_RX_U_RX   \
 	| NORMAL_OUTER_INNER_NON_CACHEABLE_SHAREABLE)})
 
+#define K_MEM_PARTITION_P_RW_U_NA_DEVICE_SHAREABLE ((k_mem_partition_attr_t) \
```
This is fine. I wonder if you can add the declaration for the ARMv8-M architecture as well. 
Optional.

reloZid replied:
No longer needed, therefore removed.

ioannisg commented on include/arch/arm/aarch32/cortex_m/mpu/arm_mpu_v7m.h:
```diff
@@ -262,6 +262,10 @@ typedef struct {
 	| NORMAL_OUTER_INNER_NON_CACHEABLE_SHAREABLE)})
 
+#define K_MEM_PARTITION_P_RW_U_NA_DEVICE_SHAREABLE ((k_mem_partition_attr_t) \
+	{(_K_MEM_PARTITION_P_RW_U_NA \
```
nit: see if you could have it in a single line

reloZid replied:
No longer needed, therefore removed.

ioannisg commented on arch/arm/core/aarch32/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -92,6 +92,18 @@ static const struct z_arm_mpu_partition static_regions[] = {
 			.attr = K_MEM_PARTITION_P_RO_U_NA,
 		},
 #endif /* !CONFIG_MULTITHREADING && CONFIG_MPU_STACK_GUARD */
+#if defined(CONFIG_ETH_STM32_HAL_USE_SRAM3)
```
This section is 
- SoC specific, not feature specific, and,
- outside the image boundaries (image_ram_start, image_ram_end)

So I think we should try to define this MPU region at boot. See soc/arm/common/arm_mpu_regions.c. As discussed in the dev-review meeting there is a Kconfig option that allows to override the default SoC MPU region definition, and add a SoC-specific one that includes _this_ region as well.

reloZid replied:
Thanks for this hint! Added SoC specific MPU region definitions.

ioannisg reviewed (changes requested):
Looks acceptable to me, but I would insist on trying using DT for RAM region definition and use the DT macros in the linker script and other locations. And use the SOC_SPECIFIC MPU region definition option and define the MPU region at the SoC level (not in the arm mpu driver)

Future:
I would really like to clean this up; in particular, the soc could add hooks to the mpu driver at run time, to define **_static_** MPU partitions that are SoC-specific, not feature-specific.

reloZid commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -78,6 +78,13 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #define CACHE
 #endif
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X) && DT_NODE_HAS_STATUS(DT_NODELABEL(sram3), okay)
```
Should there still be a Kconfig option to enable this for the driver?

ioannisg replied:
@erwango ^^

reloZid commented on dts/arm/st/h7/stm32h747.dtsi:
```diff
@@ -5,3 +5,41 @@
+
+/ {
+	/*
+	* The RAM memories placed here can be used by both cores M4/M7
```
Aligned this with the H745 device tree which should have the same memory architecture. Anyone can confirm this is correct?

erwango replied:
Please have a check to https://github.com/zephyrproject-rtos/zephyr/pull/30530.
I'm reworking that part  a bit.

Nukersson replied:
> Aligned this with the H745 device tree which should have the same memory architecture. Anyone can confirm this is correct?

Yes. They have same memory architecture. They also have same ReferenceManual

reloZid commented on soc/arm/st_stm32/stm32h7/mpu_regions.c:
```diff
@@ -0,0 +1,66 @@
+#include <devicetree.h>
+#include <arch/arm/aarch32/cortex_m/mpu/arm_mpu.h>
+
+/* Flash Region Definitions */
```
Copied this from `soc/arm/common/cortex_m/arm_mpu_mem_cfg.h`. Can we share this somehow?

erwango replied:
Isn't it possible to include `soc/arm/common/cortex_m/arm_mpu_mem_cfg.h` directly ?

reloZid replied:
Included it directly. Wasn't sure because it seems only possible using a relative path:

```
#include "../../common/cortex_m/arm_mpu_mem_cfg.h"
```

reloZid commented on dts/arm/st/h7/stm32h745.dtsi:
```diff
@@ -21,11 +21,23 @@
 
 	/* System data RAM accessible over AHB bus: SRAM1 in D2 domain */
 	sram1: memory@30000000 {
-		reg = <0x30000000 DT_SIZE_K(288)>;
```
This covered the whole D2 RAM: SRAM1, SRAM2 and SRAM3

I hope it is fine to separate it.

Nukersson replied:
I think, it is okay for the first time. Most important thing here is that, to know, how to generate linker sections with MPU configuration. Therefore it is not important, which and how much sram were used. Let it as it stays here. In the next step, 9t will be improved.

erwango commented:
@reloZid it might be interested for you to rebase on top of latest master, following the merge of https://github.com/zephyrproject-rtos/zephyr/pull/30530.
Also, please remove the initial commits, so we only have to review the last version of the proposed solution. This should help in the review process

reloZid commented:
Rebased and adapted to pass all checks. Please note that I only tested the fix manually using the dumb HTTP server sample on a Nucleo H743. And in general I don't know the MPU well enough to tell whether this solution will work in any case.

Nukersson commented on soc/arm/st_stm32/stm32h7/sections.ld:
```diff
@@ -0,0 +1,17 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(sram3), okay) && DT_NODE_HAS_STATUS(DT_NODELABEL(mac), okay)
```
I would prefer to name this region `no_cache_no_mpu_sram` and add it accordingly to KConfig of stm32

Nukersson commented on soc/arm/st_stm32/stm32h7/sections.ld:
```diff
@@ -0,0 +1,17 @@
+
+SECTION_DATA_PROLOGUE(eth_stm32,(NOLOAD),)
+{
+    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(sram3)));
```
Same here please: `no_cache_no_mpu_sram`

Nukersson commented on soc/arm/st_stm32/stm32h7/sections.ld:
```diff
@@ -0,0 +1,17 @@
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(sram3), okay) && DT_NODE_HAS_STATUS(DT_NODELABEL(mac), okay)
+
+SECTION_DATA_PROLOGUE(eth_stm32,(NOLOAD),)
```
Same here: `no_cache_no_mpu_sram`

KozhinovAlexander reviewed (commented):
@erwango What if we add some common `non_mpu_sram` sram region and depending on it existence in KConfig file configure the MPU accordingly. This region will still be defined in linker file always, but disabled by prepocessor's `#if defined(non_mpu_sram)` directive.

Nukersson commented on dts/arm/st/h7/stm32h743.dtsi:
```diff
@@ -15,10 +15,35 @@
+	/* System data RAM accessible over AHB bus: SRAM3 in D2 domain */
+	sram3: memory@30040000 {
+		compatible = "mmio-sram";
+		reg = <0x30040000 DT_SIZE_K(32)>;
```
```suggestion
		reg = <0x30040000 DT_SIZE_K(8)>;
```

Nukersson commented on dts/arm/st/h7/stm32h743.dtsi:
```diff
@@ -15,10 +15,35 @@
+		reg = <0x30020000 DT_SIZE_K(128)>;
+	};
+
+	/* System data RAM accessible over AHB bus: SRAM3 in D2 domain */
+	sram3: memory@30040000 {
+		compatible = "mmio-sram";
+		reg = <0x30040000 DT_SIZE_K(32)>;
+	};
```
```suggestion
	/* System data RAM accessible over AHB bus: SRAM3 in D2 domain */
	sram3: memory@30040000 {
		compatible = "mmio-sram";
		reg = <0x30040000 DT_SIZE_K(24)>;
	};
	
	/* System data RAM accessible over AHB bus with MPU disabled: SRAM3 in D2 domain */
	non_mpu_sram: memory@30046000 {
		compatible = "mmio-sram";
		reg = <0x30046000 DT_SIZE_K(8)>;
	};
```

gmarull replied:
I don't think DT definitions need to be related with MPU settings. MPU for a certain memory region may or may not be configured depending on the application or driver needs.

Nukersson replied:
> I don't think DT definitions need to be related with MPU settings. MPU for a certain memory region may or may not be configured depending on the application or driver needs.

@gmarull I agree with it. My idea here is to define or not define non_mpu_sram and get this memory region configured or not. If this memory region does not exist in DT, then everything works like usual. In other case non_mpu_sram will be accordingly configured.

gmarull replied:
> @gmarull I agree with it. My idea here is to define or not define non_mpu_sram and get this memory region configured or not. If this memory region does not exist in DT, then everything works like usual. In other case non_mpu_sram will be accordingly configured.

Applications that may require full SRAM3 (ie 32K) will not work as usual, as you're suggesting to split it.

Nukersson replied:
> > @gmarull I agree with it. My idea here is to define or not define non_mpu_sram and get this memory region configured or not. If this memory region does not exist in DT, then everything works like usual. In other case non_mpu_sram will be accordingly configured.
> 
> Applications that may require full SRAM3 (ie 32K) will not work as usual, as you're suggesting to split it. 

@gmarull Sure, if non_mpu_sram is enabled. But we can properly document it. Otherwise, your ideas on MPU disabled sram region?

gmarull replied:
It's up to the application how MPU is configured, not SRAM descriptions in DT.

reloZid replied:
I also don't think we should split the SRAM3 in the DT. If we want to reserve only 8K for the Ethernet driver such that the rest can be used by other drivers, we can adapt the MPU region sizes and enforce our section to be exactly 8K in the linker script. Then other drivers can fill up the rest of SRAM3 with their own sections. I thought it would be okay for now as is because there is currently no other use case for SRAM3.

Nukersson replied:
Okay. Thanks for your ideas.

Nukersson commented on dts/arm/st/h7/stm32h745.dtsi:
```diff
@@ -35,11 +35,23 @@
+		reg = <0x30020000 DT_SIZE_K(128)>;
+	};
+
+	/* System data RAM accessible over AHB bus: SRAM3 in D2 domain */
+	sram3: memory@30040000 {
+		compatible = "mmio-sram";
+		reg = <0x30040000 DT_SIZE_K(32)>;
+	};
```
```suggestion
	/* System data RAM accessible over AHB bus: SRAM3 in D2 domain */
	sram3: memory@30040000 {
		compatible = "mmio-sram";
		reg = <0x30040000 DT_SIZE_K(24)>;
	};
	
	/* System data RAM accessible over AHB bus with MPU disabled: SRAM3 in D2 domain */
	non_mpu_sram: memory@30046000 {
		compatible = "mmio-sram";
		reg = <0x30046000 DT_SIZE_K(8)>;
	};
```

Nukersson replied:
No more actual. This part looks good to me.

Nukersson commented on dts/arm/st/h7/stm32h750.dtsi:
```diff
@@ -7,10 +7,34 @@
+		reg = <0x30020000 DT_SIZE_K(128)>;
+	};
+
+	/* System data RAM accessible over AHB bus: SRAM3 in D2 domain */
+	sram3: memory@30040000 {
+		compatible = "mmio-sram";
+		reg = <0x30040000 DT_SIZE_K(32)>;
+	};
```
```suggestion
	/* System data RAM accessible over AHB bus: SRAM3 in D2 domain */
	sram3: memory@30040000 {
		compatible = "mmio-sram";
		reg = <0x30040000 DT_SIZE_K(24)>;
	};
	
	/* System data RAM accessible over AHB bus with MPU disabled: SRAM3 in D2 domain */
	non_mpu_sram: memory@30046000 {
		compatible = "mmio-sram";
		reg = <0x30046000 DT_SIZE_K(8)>;
	};
```

Nukersson replied:
No more actual. This part looks good to me.

Nukersson commented on include/arch/arm/aarch32/cortex_m/scripts/linker.ld:
```diff
@@ -106,6 +106,9 @@ MEMORY
     SRAM1                 (rw) : ORIGIN = RAM1_ADDR, LENGTH = RAM1_SIZE
     SRAM2                 (rw) : ORIGIN = RAM2_ADDR, LENGTH = RAM2_SIZE
 #endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(sram3), okay)
+    SRAM3                 (rw) : ORIGIN = DT_REG_ADDR(DT_NODELABEL(sram3)), LENGTH = DT_REG_SIZE(DT_NODELABEL(sram3))
+#endif
```
```suggestion
#if DT_NODE_HAS_STATUS(DT_NODELABEL(non_mpu_sram), okay)
    NON_MPU_SRAM                 (rw) : ORIGIN = DT_REG_ADDR(DT_NODELABEL(non_mpu_sram)), LENGTH = DT_REG_SIZE(DT_NODELABEL(non_mpu_sram))
#endif
```

Nukersson replied:
No more actual. This part looks good to me.

Nukersson commented on soc/arm/st_stm32/stm32h7/mpu_regions.c:
```diff
@@ -0,0 +1,79 @@
+					 REGION_FLASH_ATTR(REGION_FLASH_SIZE)),
+	MPU_REGION_ENTRY("SRAM", CONFIG_SRAM_BASE_ADDRESS,
+					 REGION_RAM_ATTR(REGION_SRAM_SIZE)),
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(sram3), okay) && \
+		DT_NODE_HAS_STATUS(DT_NODELABEL(mac), okay)
+	MPU_REGION_ENTRY("SRAM3_ETH_BUF",
+					 DT_REG_ADDR(DT_NODELABEL(sram3)),
+					 REGION_NOCACHE_ATTR(REGION_32K)),
+	MPU_REGION_ENTRY("SRAM3_ETH_DESC",
+					 DT_REG_ADDR(DT_NODELABEL(sram3)),
+					 REGION_PPB_ATTR(REGION_256B)),
+#endif
```
```suggestion
#if DT_NODE_HAS_STATUS(DT_NODELABEL(non_mpu_sram), okay) && \
		DT_NODE_HAS_STATUS(DT_NODELABEL(mac), okay)
	MPU_REGION_ENTRY("NON_MPU_SRAM_ETH_BUF",
					 DT_REG_ADDR(DT_NODELABEL(non_mpu_sram)),
					 REGION_NOCACHE_ATTR(REGION_SRAM_SIZE)),
	MPU_REGION_ENTRY("NON_MPU_SRAM_ETH_DESC",
					 DT_REG_ADDR(DT_NODELABEL(non_mpu_sram)),
					 REGION_PPB_ATTR(REGION_256B)),
#endif
```
Not sure if `REGION_SRAM_SIZE` correct here, maybe `REGION_8K` would be okay for the first time.

reloZid replied:
Yes I think that should be `REGION_8K`, because `REGION_SRAM_SIZE` resolves to the size of the main system memory.

reloZid replied:
To be honest, I think the name `non_mpu_sram` is a bit misleading, because the region uses a special MPU configuration, but the MPU is not disabled. For example the MPU still makes sure that the data in this region is not executable.

Nukersson replied:
Let stay with SRAM3

Nukersson commented on soc/arm/st_stm32/stm32h7/sections.ld:
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2020 Mario Jaun
```
@reloZid You're welcome to put your copyright just right down here.

reloZid replied:
Actually this is my copyright ;-) Sorry I'm not very familiar with legal stuff. If I create a new file, who do I have to mention in the copyright? And under which circumstances shall I append my copyright to an existing file?

Nukersson replied:
> Actually this is my copyright ;-) Sorry I'm not very familiar with legal stuff. If I create a new file, who do I have to mention in the copyright?

You don't have to mention anybody, if it is your choice. But you can also put just your nickname and date, what should be okay.

> And under which circumstances shall I append my copyright to an existing file?

If I am right, the CI expects copyright. But you can put anything there. Still, some meaningful copyright even with your nickname would be okay.

erwango replied:
There is no check on this, but I don't think nick names are ok (copyright is "serious stuff")

Nukersson commented on soc/arm/st_stm32/stm32h7/sections.ld:
```diff
@@ -0,0 +1,17 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(sram3), okay) && DT_NODE_HAS_STATUS(DT_NODELABEL(mac), okay)
+
+SECTION_DATA_PROLOGUE(eth_stm32,(NOLOAD),)
+{
+    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(sram3)));
+    *(.eth_stm32_desc)
+    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(sram3))) + 256;
+    *(.eth_stm32_buf)
+} GROUP_DATA_LINK_IN(SRAM3, SRAM3)
+
+#endif
```
```suggestion
#if DT_NODE_HAS_STATUS(DT_NODELABEL(non_mpu_sram), okay) && DT_NODE_HAS_STATUS(DT_NODELABEL(mac), okay)

SECTION_DATA_PROLOGUE(eth_stm32,(NOLOAD),)
{
    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(non_mpu_sram)));
    *(.eth_stm32_desc)
    . = ABSOLUTE(DT_REG_ADDR(DT_NODELABEL(non_mpu_sram))) + 256;
    *(.eth_stm32_buf)
} GROUP_DATA_LINK_IN(NON_MPU_SRAM, NON_MPU_SRAM)

#endif
```

Nukersson replied:
No more actual. This part looks good to me.

KozhinovAlexander reviewed (changes requested):
@reloZid Please take a look again.

Nukersson commented:
As a first step towards specific MPU regions for stm32 LGTM.

Nukersson commented:
@erwango In #29915 it is mentioned, that semaphore timeout bug happens only with `CONFIG_DEBUG` setting. Thus, I would propose to add here mentioned MPU configuration only for debug compiled project. Otherwise it works fine. 
My another opinion is, to ask MPU responsible developers, of such setting is meaningful for all regions in debug mode.

erwango commented:
> My another opinion is, to ask MPU responsible developers, of such setting is meaningful for all regions in debug mode.

@ioannisg ?

reloZid commented:
Did some final cleanups:

- Included `arm_mpu_mem_cfg.h` in `mpu_regions.c` to avoid duplication
- Adapted linker sections to reserve exactly 16K of SRAM3 for our Ethernet buffers (minimum MPU region size we need), such that the remaining 16K are free for other uses
- Rebased branch onto master

Please note that I could only test the PR manually using the dumb_http_server sample on an Nucleo H743.

carlescufi commented:
@erwango and @ioannisg can you please re-review?

carlescufi commented:
@gmarull are you happy with this as-is?

gmarull reviewed (changes requested):
I think that for now it is okay. Only change request: please split changes into multiple commits.

It is clear though that Zephyr needs a better way to define custom MPU regions (e.g. see https://github.com/u-boot/u-boot/blob/master/arch/arm/mach-stm32/soc.c) and so get rid of `CPU_HAS_CUSTOM_FIXED_SOC_MPU_REGIONS`.

ioannisg commented:
will take another look asap after 2.5.0 is out

ioannisg commented on include/arch/arm/aarch32/cortex_m/mpu/arm_mpu_v7m.h:
```diff
@@ -115,6 +115,11 @@
 }
+#define REGION_NOCACHE_ATTR(size) \
+{ \
+	(NORMAL_OUTER_INNER_NON_CACHEABLE_SHAREABLE | \
```
@reloZid what is the reason this needs to be declared as Sharable?

ioannisg replied:
I also wonder, whether we could name it REGION_RAM_NOCACHE_ATTR, since we only seem to need this for RAM region configurations, that is, make the name more explicitly reflect the usage.

reloZid replied:
@ionnisg Adopted this from an example from ST. But also seems to work as non-shareable. Applied both your suggestions.

mjaun committed:
dts: arm: st: h7: define SRAM1-4

Define SRAM1 to SRAM4 memory areas according to the physical memory
organization of the chips.

Signed-off-by: Mario Jaun <mario.jaun@gmail.com>

mjaun committed:
linker: arm: add SRAM3/4 linker section

Add a linker section for SRAM3/4 if it is enabled in the device tree.

Signed-off-by: Mario Jaun <mario.jaun@gmail.com>

mjaun committed:
arch: arm: add define REGION_RAM_NOCACHE_ATTR

Add a define for non cachaeable RAM for MPU region definitions.

Signed-off-by: Mario Jaun <mario.jaun@gmail.com>

mjaun committed:
drivers: ethernet: stm32: SRAM3 / MPU configuration

Fixes #29915.

Implements the memory layout and MPU configuration for Ethernet buffers
for STM32H7 controllers as recommended by ST. 16 KB of SRAM3 are
are reserved for this. The first 256 B are for the RX/TX descriptors and
configured as strongly ordered, shareable memory. The rest is for RX/TX
buffers and configured as non cacheable memory. This configuration is
automatically applied for H7 chips if the SRAM3 memory is enabled in the
device tree.

Signed-off-by: Mario Jaun <mario.jaun@gmail.com>

reloZid commented:
> 
> 
> I think that for now it is okay. Only change request: please split changes into multiple commits.
> 
> It is clear though that Zephyr needs a better way to define custom MPU regions (e.g. see https://github.com/u-boot/u-boot/blob/master/arch/arm/mach-stm32/soc.c) and so get rid of `CPU_HAS_CUSTOM_FIXED_SOC_MPU_REGIONS`.

@gmarull Splitted the PR into four commits:

* Update device tree files
* Define SRAM3 linker section
* Add REGION_RAM_NOCACHE_ATTR define
* Adapt STM32 ETH driver

erwango commented:
@reloZid would you mind rebasing ?

reloZid commented:
@erwango Sorry for the late reaction. I rebased the changes.

erwango commented:
@ioannisg can you have a new look ?
